kythe/rust/fuchsia_extractor/README.md

```diff
@@ -1,9 +1,173 @@
+an ideal Rust extractor can use.  As work is underway to improve both of these
+two issues, we decided to go with a pragmatic approach first, and once the
+libraries are available, we can update the indexer and the extractor to match.
 
 [fx]: https://www.fuchsia.dev
+[gn]: https://gn.googlesource.com/gn/
 
-Fuchsia build system is a bit specific, and is custom built on top of [GN][gn].
+Technically, this extractor produces `kzip` files as any other extractor does,
+and the `kzip` file specification is well known.  However, in practice, the
+kzip specification is open for interpretation (for example, what if you need
+to inject a specific extra file to a compilation unit?).  That in turn can
+lead to mismatches in the interpretations that would break the ingestion
+pipeline.
+
+This document is an attempt to lay out any such assumptions for
+`fuchsia_extractor` going forward.
+
+# Specifics of Fuchsia compilation
+
+Fuchsia is compiled with [GN][gn].  However, due to the size of the code base,
+the GN generator scripts are heavily customized and templatized in a specific
+way, so it could be said that Fuchsia uses a custom purpose-built build system.
+
+Fuchsia source code is spread across many repositories, some of which can be
+seen at https://fuchsia.googlesource.com.  When the source code is checked out,
+a custom utility `fx jiri` is used (which is to say, a custom wrapper over the
+`jiri` program) to check out all the repositories into its appropriate place
+underneath a common source root directory, which we will call `$FUCHSIA_DIR`
+after the environment variable that is usually used to refer to it in scripts.
+
+Fuchsia code is compiled in a separate "output" directory.  However, this
+directory is created underneath `$FUCHSIA_DIR`, and each build configuration,
+for a "product-dot-board" combination such as `terminal.x64`, gets a separate
+build output directory.  The build configuration `terminal.x64` gets an output
+directory in `$FUCHSIA_DIR/out/terminal.x64`, though this last path component
+can be given a different name by the developer.  Let us call this directory
+`$FUCHSIA_OUT` for short.
+
+Multiple top-level directories exist in `$FUCHSIA_DIR` and `$FUCHSIA_OUT`.  The
+"source" corpus is at `../../` relative to the build directory.  FIDL generated
+code is at `./fidling`, Banjo generated code is at `./banjoing` and so on.
+Multiple roots are used to satisfy the spec requirement for `VName`, which
+requires each file path to be relative to its root, *and* not use any "parent"
+references.
+
+When filling out compilation unit data, we need to ensure that when we refer to
+files, their names are expressed relative to some of those directories, and
+then to correct ones.  Furthermore, sometimes relative paths are needed with
+respect to the root directory of the repository tree, and sometimes with
+respect to the build directory.  For example, generated code lives in
```

Comment:
The only paths which should be relative to the repository are those in the VName.  Every other path should be relative to the working_directory (or absolute if it doesn't live under the build dir), which should be the build directory.   Note that the build directory doesn't have to equal the `working_directory` but is often easier to reason about when that is the case.